### PR TITLE
Add AES-IGE block cipher

### PR DESCRIPTION
**Pull request by watzon**
This PR adds support for the AES-IGE block cipher mode. I've tried to make sure thing are done the Zig way with tests and such, but this is really my first code related PR so I'm sure improvements can be made. Specifically I'm not sure about the `std.mem.copy` calls, but I wasn't able to figure out how to have those slices be the correct type otherwise.

References #5763

### CODE REVIEW

File: lib/std/crypto/aes.zig [hunk 1]
```diff
@@ -215,6 +240,29 @@ fn AESDecrypt(comptime keysize: usize) type {
         pub fn decrypt(ctx: Self, dst: []u8, src: []const u8) void {
             decryptBlock(ctx.dec[0..], dst, src);
         }
+
+        pub fn ige(ctx: Self, dst: []u8, src: []const u8, iv: [(keysize / 8) * 2]u8) void {
+            std.debug.assert(src.len >= dst.len);
```
Comment (daurnimator) [inline lib/std/crypto/aes.zig:245]: should this be `==`?

File: lib/std/crypto/aes.zig [hunk 2]
```diff
@@ -215,6 +240,29 @@ fn AESDecrypt(comptime keysize: usize) type {
...
+            var m: [b]u8 = undefined;
+
+            std.mem.copy(u8, &c, iv[0..b]);
+            std.mem.copy(u8, &m, iv[b..]);
+
+            var o: usize = 0;
+            while (o < src.len) : (o += b) {
+                const t = src[o..o+b];
+
+                _ = xorBytes(dst[o..o+b], src[o..o+b], &m);
```
Comment (daurnimator) [inline lib/std/crypto/aes.zig:258]: Use `t` you created above?

File: lib/std/crypto/aes.zig [hunk 3]
```diff
@@ -215,6 +240,29 @@ fn AESDecrypt(comptime keysize: usize) type {
...
+            const b = keysize / 8;
+            var c: [b]u8 = undefined;
+            var m: [b]u8 = undefined;
+
+            std.mem.copy(u8, &c, iv[0..b]);
+            std.mem.copy(u8, &m, iv[b..]);
+
+            var o: usize = 0;
+            while (o < src.len) : (o += b) {
+                const t = src[o..o+b];
```
Comment (daurnimator) [inline lib/std/crypto/aes.zig:256]: Write as `const t = src[o..][0..b]` instead to get a comptime known length slice (pointer to array)
Add `const d = dst[o..][0..b]` as well?

File: lib/std/crypto/aes.zig [hunk 4]
```diff
@@ -215,6 +240,29 @@ fn AESDecrypt(comptime keysize: usize) type {
...
+
+                _ = xorBytes(dst[o..o+b], src[o..o+b], &m);
+                ctx.decrypt(dst[o..o+b], dst[o..o+b]);
+                _ = xorBytes(dst[o..o+b], dst[o..o+b], &c);
+
+                std.mem.copy(u8, &m, dst[o..o+b]);
+                std.mem.copy(u8, &c, t);
+            }
+        }
```
Comment (daurnimator) [inline lib/std/crypto/aes.zig:265]: what happens if `src.len` is not divisible by `b`?
either need handling for the tail; or an assert + doc comment.

File: lib/std/crypto/aes.zig (hunk 1 above)
Comment (watzon) [reply lib/std/crypto/aes.zig:245]: I actually think it's the "encrypt" variant that's wrong. Technically I believe you should be able to have a `dst` that's bigger than `src`, but not vice-versa.

Comment (watzon) [reply lib/std/crypto/aes.zig:245]: Actually this is wrong too because the conditions should be reversed

**Comment by watzon** [comment]
Ok things have been refactored a bit. Tests still pass, but I added `blocklen` as a top level constant for `AES`, `AESEncrypt`, and `AESDecrypt`, modified the assertions so that `src.len` has to be divisible by `blocklen` and `dst.len >= src.len`, and added the `t` and `d` constants to the while loops to avoid calling `src[o..][0..blocklen]` and `dst[o..][0..blocklen]` all over the place.

**Comment by daurnimator** [comment]
The code looks reasonable to me.

Might be good for you to rebase to get rid of the merge and your intermediate work.

**Comment by watzon** [comment]
IGE itself is really only used for Telegram, but it's one of many block cipher implementations I'm planning on adding. As I see it, if the goal of Zig is to be able to stand independent of C, it makes sense to have as strong a crypto implementation as possible. Right now the AES implementation is really lackluster and missing a lot of the block ciphers which are used pretty commonly in many different applications, and the one cipher that is included, CTR, is even missing a way to decrypt the encrypted payload.

**Comment by daurnimator** [comment]
> IGE itself is really only used for Telegram,

Isn't it only used in the old deprecated telegram API?

**Comment by watzon** [comment]
No it's used currently in the MTProto v2 spec. Afaik there's actually nothing wrong with IGE, other than maybe efficiency. I believe the majority of the reason it isn't used is because a) no one has heard about it, and b) OpenSSL deprecated it in their library because their implementation was broken.

**Comment by jedisct1** [comment]
Having a rich standard library is absolutely great.

However, there is always a balance to find between adding features so that applications don't need much external dependencies, and keeping the standard library lean in order to ensure that everything in here remains stable, well-maintained and actually used.

IGE can be okay, if used correctly. But similar to CBC, it didn't age well. It is slow, cannot be parallelized, and requires randomness. Integrity protection by adding an extra block is not as safe as advertised. It is not part of any standard suite, so you have to choose your own MAC. Plus, IGE implementations, unlike CBC, are scarce and diverge on how the initial state should be initialized.  

It has virtually no real-world deployment besides MTProto, and it is very unlikely to get any more traction in the future as modern modes have way more compelling properties.

In this context, maybe AES-IGE would be better maintained inside a Telegram client library than in Zig's standard library. This ensures that the variant implemented is totally compatible with MTProto and that it will be well-tested by actual applications using it. Of course, this can be revisited if AES-IGE gains more popularity, but until then, something specifically designed and optimized for MTProto, that is part of an MTProto implementation, probably makes more sense.

**Comment by andrewrk** [comment]
Thanks for sending this patch upstream.

> It might be something that would go better in an external library, but it might be worth it to discuss what the determining factor should be for whether or not a cipher belongs in the standard library.

Honest answer to this is I'm not sure yet. I wrote a little bit about it here: https://github.com/ziglang/zig/issues/5763#issuecomment-691258518

However, given that we started bumping against (temporary/stage1) resource limits of compiling the standard library tests, and started the [std-lib-orphanage](https://github.com/ziglang/std-lib-orphanage) project, I think it makes sense to take a "when in doubt, keep it external" approach.

I'll put some thought into a more official and clear policy on what goes in and what doesn't. Until then I am closing this.

**Pull request closed without merging**